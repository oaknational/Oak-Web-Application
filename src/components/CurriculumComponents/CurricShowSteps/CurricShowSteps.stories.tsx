import { Meta, StoryObj } from "@storybook/react";

import { CurricShowSteps as Component } from ".";

const meta: Meta<typeof Component> = {
  component: Component,
  tags: ["autodocs"],
  argTypes: {
    numberOfSteps: {
      type: "number",
      control: {
        type: "range",
        min: 2,
        max: 5,
      },
    },
    currentStep: {
      type: "number",
      control: "number",
    },
  },
  parameters: {
    controls: {
      include: ["numberOfSteps", "currentStep"],
    },
  },
  render: (args) => <Component {...args} />,
};

export default meta;
type Story = StoryObj<typeof Component>;

export const Default: Story = {
  args: {
    currentStep: 2,
    numberOfSteps: 4,
  },
};
