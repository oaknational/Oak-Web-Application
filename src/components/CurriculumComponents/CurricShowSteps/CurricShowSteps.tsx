import {
  OakBox,
  OakColorToken,
  OakFlex,
  OakFlexProps,
  OakIcon,
  OakIconName,
} from "@oaknational/oak-components";
import { Fragment } from "react";

const DOT_SIZE = 44;

function StepLine({ $borderStyle }: Pick<OakFlexProps, "$borderStyle">) {
  return (
    <OakFlex
      $borderStyle={$borderStyle}
      $ba="border-solid-none"
      $bb={"border-solid-m"}
      $flexGrow={1}
    />
  );
}

type DotWithWingsProps = {
  isFirst: boolean;
  isLast: boolean;
  iconName?: OakIconName;
  $background?: OakColorToken;
};
function DotWithWings({
  isFirst,
  isLast,
  iconName,
  $background,
}: DotWithWingsProps) {
  return (
    <OakFlex
      style={{ width: DOT_SIZE, height: DOT_SIZE }}
      $alignItems={"center"}
      $justifyContent={"center"}
    >
      <StepLine $borderStyle={isFirst ? "none" : "solid"} />
      <OakFlex
        $background={$background}
        $width={"all-spacing-7"}
        $height={"all-spacing-7"}
        $borderRadius={"border-radius-circle"}
        $alignItems={"center"}
        $justifyContent={"center"}
        $borderStyle={"solid"}
        $ba="border-solid-m"
        $borderColor={"black"}
      >
        {iconName && (
          <OakIcon
            iconName={iconName}
            $colorFilter="white"
            $width={"all-spacing-6"}
            $height={"all-spacing-6"}
          />
        )}
      </OakFlex>
      <StepLine $borderStyle={isLast ? "none" : "solid"} />
    </OakFlex>
  );
}

type CurricShowStepsProps = {
  numberOfSteps: number;
  currentStep: number;
};
export function CurricShowSteps({
  numberOfSteps,
  currentStep,
}: CurricShowStepsProps) {
  return (
    <div>
      <OakBox
        style={{ width: "100%", position: "relative", overflow: "hidden" }}
        $pb={"inner-padding-xl"}
      >
        <div style={{ display: "flex", alignItems: "center" }}>
          {Array(numberOfSteps)
            .fill(true)
            .map((_, index) => {
              const isLast = index === numberOfSteps - 1;
              const isFirst = index === 0;
              const isCurrent = index === currentStep;
              const isDone = index < currentStep;

              return (
                <Fragment key={index}>
                  {isDone && (
                    <DotWithWings
                      isFirst={isFirst}
                      isLast={isLast}
                      iconName="tick"
                      $background="black"
                    />
                  )}
                  {!isCurrent && !isDone && (
                    <DotWithWings
                      isFirst={isFirst}
                      isLast={isLast}
                      $background="white"
                    />
                  )}
                  {isCurrent && (
                    <OakFlex
                      $borderStyle={"solid"}
                      $borderColor={"black"}
                      $borderRadius={"border-radius-circle"}
                      $alignItems={"center"}
                      $justifyContent={"center"}
                      $ba="border-solid-m"
                      $position={"relative"}
                      style={{ width: DOT_SIZE, height: DOT_SIZE }}
                    >
                      <OakFlex
                        $background={"black"}
                        $width={"all-spacing-7"}
                        $height={"all-spacing-7"}
                        $borderRadius={"border-radius-circle"}
                        color="white"
                        $alignItems={"center"}
                        $justifyContent={"center"}
                       />
                      <OakBox
                        $position="absolute"
                        $whiteSpace={"nowrap"}
                        data-testid="step-text"
                        style={{
                          [isLast ? "right" : "left"]: 0,
                          top: DOT_SIZE,
                        }}
                      >
                        Step {currentStep + 1} of {numberOfSteps}
                      </OakBox>
                    </OakFlex>
                  )}
                  {!isLast && <StepLine />}
                </Fragment>
              );
            })}
        </div>
      </OakBox>
    </div>
  );
}
