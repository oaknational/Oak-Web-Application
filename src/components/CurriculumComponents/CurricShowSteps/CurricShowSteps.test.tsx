import { CurricShowSteps } from "./CurricShowSteps";

import renderWithTheme from "@/__tests__/__helpers__/renderWithTheme";

describe("CurricShowSteps", () => {
  test("basic usage", () => {
    const { baseElement, getByTestId } = renderWithTheme(
      <CurricShowSteps numberOfSteps={4} currentStep={2} />,
    );
    expect(baseElement).toMatchSnapshot();
    expect(getByTestId("step-text")).toHaveTextContent("Step 3 of 4");
  });
});
