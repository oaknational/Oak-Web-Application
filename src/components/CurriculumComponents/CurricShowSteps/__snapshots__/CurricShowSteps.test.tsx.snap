// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurricShowSteps basic usage 1`] = `
<body>
  <div>
    <div>
      <div
        class="sc-fqkvVR hrLDNc"
        style="width: 100%; position: relative; overflow: hidden;"
      >
        <div
          style="display: flex; align-items: center;"
        >
          <div
            class="sc-fqkvVR sc-kAyceB fgBWrW kyta-DD"
            style="width: 44px; height: 44px;"
          >
            <div
              class="sc-fqkvVR sc-kAyceB ibvSUR bAFsaQ"
            />
            <div
              class="sc-fqkvVR sc-kAyceB hvDzQc kyta-DD"
            >
              <div
                class="sc-fqkvVR kbnxKi"
              >
                <img
                  alt=""
                  class="sc-dcJsrY bDcxkE"
                  data-nimg="fill"
                  decoding="async"
                  loading="lazy"
                  src="https://next_public_oak_assets_host/NEXT_PUBLIC_OAK_ASSETS_PATH/v1699954310/icons/efd3esaor6zqk7seh6kt.svg"
                  style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
                />
              </div>
            </div>
            <div
              class="sc-fqkvVR sc-kAyceB jkxpdi bAFsaQ"
            />
          </div>
          <div
            class="sc-fqkvVR sc-kAyceB cdDatI bAFsaQ"
          />
          <div
            class="sc-fqkvVR sc-kAyceB fgBWrW kyta-DD"
            style="width: 44px; height: 44px;"
          >
            <div
              class="sc-fqkvVR sc-kAyceB jkxpdi bAFsaQ"
            />
            <div
              class="sc-fqkvVR sc-kAyceB hvDzQc kyta-DD"
            >
              <div
                class="sc-fqkvVR kbnxKi"
              >
                <img
                  alt=""
                  class="sc-dcJsrY bDcxkE"
                  data-nimg="fill"
                  decoding="async"
                  loading="lazy"
                  src="https://next_public_oak_assets_host/NEXT_PUBLIC_OAK_ASSETS_PATH/v1699954310/icons/efd3esaor6zqk7seh6kt.svg"
                  style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
                />
              </div>
            </div>
            <div
              class="sc-fqkvVR sc-kAyceB jkxpdi bAFsaQ"
            />
          </div>
          <div
            class="sc-fqkvVR sc-kAyceB cdDatI bAFsaQ"
          />
          <div
            class="sc-fqkvVR sc-kAyceB ffuCSZ kyta-DD"
            style="width: 44px; height: 44px;"
          >
            <div
              class="sc-fqkvVR sc-kAyceB jAHGNa kyta-DD"
              color="white"
            />
            <div
              class="sc-fqkvVR cMqhUz"
              data-testid="step-text"
              style="left: 0px; top: 44px;"
            >
              Step 
              3
               of 
              4
            </div>
          </div>
          <div
            class="sc-fqkvVR sc-kAyceB cdDatI bAFsaQ"
          />
          <div
            class="sc-fqkvVR sc-kAyceB fgBWrW kyta-DD"
            style="width: 44px; height: 44px;"
          >
            <div
              class="sc-fqkvVR sc-kAyceB jkxpdi bAFsaQ"
            />
            <div
              class="sc-fqkvVR sc-kAyceB jGUeSQ kyta-DD"
            />
            <div
              class="sc-fqkvVR sc-kAyceB ibvSUR bAFsaQ"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
